use snmalloc as global allocator (#1848)
spiceai/spiceai: docs/decisions/001-use-snmalloc-as-global-allocator.md

# DR: Use snmalloc As Global Allocator

## Context

> Spice performs federated and accelerated queries on large datasets, requiring intensive memory allocation and deallocation, which can result in memory fragmentation.

Reviewer: ```suggestion
Spice performs federated, accelerated queries on large datasets, requiring frequent memory allocation and deallocation, leading to fragmentation.
```

> Rust uses system allocator by default which may struggle with this in some Operating Systems, leading to continual memory growth. Switching to a different allocator can help maintain a consistent memory footprint in Spice.

Reviewer: ```suggestion
Rust uses the system allocator by default, which may struggle in some Operating Systems under load, causing unbounded memory growth. Switching to a different allocator can help maintain a stable memory footprint in Spice.
```

## Assumptions

> 1. The chosen allocator resolves the issue of continual memory growth in primary operating systems.

Reviewer: ```suggestion
1. The chosen allocator resolves the issue of unbounded memory growth when Spice is under load.
```

> 2. The chosen allocator is as performant as, if not more so than, the system allocator.

Reviewer: ```suggestion
2. The chosen allocator is as fast or faster than the system allocator.
```

> 3. Spice's stability will not be impacted by the change of allocator in primary operating systems.

Reviewer: ```suggestion
3. Changing the allocator will not affect Spice’s stability.
```

## Options

1. [jemalloc/jemalloc](https://github.com/jemalloc/jemalloc)
2. [microsoft/mimalloc](https://github.com/microsoft/mimalloc)
3. [microsoft/snmalloc](https://github.com/Microsoft/snmalloc)

## First-Principles

- **Align to industry standards**

## Decision

> Spice will use snmalloc as the default allocator.

Reviewer: ```suggestion
Spice will use `snmalloc` as the default memory allocator on non-Windows systems.
```

**Why**:

> - snmalloc fixes the issue of continual memory growth under the same workload that system allocator couldn't handle. (debian bookworm)

Reviewer: ```suggestion
- With Spice running in our production environment on Debian Bookworm, switching the global allocator to `snmalloc` fixes the issue of unbounded memory growth under the same workload that system allocator failed on.
```

> - Benchmarks and resource monitoring indicate that using snmalloc in Spice yields slightly higher query throughput with Arrow acceleration and reduces 10%-20% memory usage comparing to jemalloc and mimalloc.

Reviewer: ```suggestion
- Benchmarks and resource monitoring show that using `snmalloc` in Spice with Arrow acceleration increases query throughput slightly and reduces memory usage by 10%-20% compared to `jemalloc` and `mimalloc`.
```

> - DataFusion, Spice's core dependency, mentions using snmalloc-rs as [optimized configuration](https://datafusion.apache.org/user-guide/example-usage.html).

Reviewer: ```suggestion
- DataFusion, Spice's core dependency, [recommends using `snmalloc-rs`](https://datafusion.apache.org/user-guide/example-usage.html) for optimized performance.
```

> - snmalloc is actively maintained and supported by Microsoft.

Reviewer: ```suggestion
- `snmalloc` is actively maintained and supported by Microsoft.
```

**Why not**:

> The [Rust RFC-1974](https://rust-lang.github.io/rfcs/1974-global-allocators.html#jemalloc) outlined the reason for not using jemalloc as the default allocator, which also have relevance to snmalloc.

Reviewer: ```suggestion
The [Rust RFC-1974](https://rust-lang.github.io/rfcs/1974-global-allocators.html#jemalloc) outlined reasons for not using `jemalloc` as the default allocator, which are also relevant to `snmalloc`.
```

> To address this, Spice ensures compatibility with primary operating systems through end-to-end CI testing. Users requiring alternative memory allocators can easily build custom versions from the trunk with minimal changes on `#[global_allocator]`.

Reviewer: ```suggestion
To address this, Spice ensures compatibility with supported operating systems via end-to-end CI testing. Users needing alternative memory allocators can build custom versions from trunk with minimal changes by swapping out the `#[global_allocator]`.
```

## Consequences

> - Implement snmalloc as the `global_allocator` in `spiced` `main.rs` via the [snmalloc-rs](https://crates.io/crates/snmalloc-rs) crate.

Reviewer: ```suggestion
- `snmalloc` is set as the `global_allocator` in `spiced` `main.rs` via the [snmalloc-rs](https://crates.io/crates/snmalloc-rs) crate.
```

> - Instabilities may arise as explained in **Why note**, the decision should be revisited as Spice evolves and new discoveries are made in this area.

Reviewer: ```suggestion
- Instabilities may arise as explained in **Why not**, the decision should be revisited as Spice evolves and new discoveries are made in this area.
```
